Add HTTP semantic convention stability migration guide (#612)
open-telemetry/semantic-conventions: docs/http/migration-guide.md

# HTTP semantic convention stability migration guide

Due to the significant number of modifications and the extensive user base
affected by them, existing HTTP instrumentations published by

> OpenTelemetry are required to implement a migration plan that will assist users in

Reviewer: Might be good to spell out that this period is intended to be 6 months, or some timeline here. That's only inferable when you look later on and see the need to maintain the old values for that period of time.

Author: can you clarify? do you mean something other than

> - Need to maintain (security patching at a minimum) their existing major version
  for at least six months after it starts emitting both sets of conventions.

transitioning to the stable HTTP semantic conventions.

Specifically, when existing HTTP instrumentations published by OpenTelemetry are
updated to the stable HTTP semantic conventions, they:

> - Need to introduce an environment variable `OTEL_SEMCONV_STABILITY_OPT_IN` in

Reviewer 2: Is `Need` the proper "level" here?  I thought we used `SHOULD` for the env var.

Reviewer: Yep, much as I would like this to be a requirement, I was also under the impression that it was optional.

Author: ```suggestion
- SHOULD introduce an environment variable `OTEL_SEMCONV_STABILITY_OPT_IN` in
```

  their existing major version, which accepts:
  - `http` - emit the stable HTTP and networking conventions, and stop emitting
    the old HTTP and networking conventions that the instrumentation emitted
    previously.
  - `http/dup` - emit both the old and the stable HTTP and networking
    conventions, allowing for a phased rollout of the stable semantic
    conventions.
  - The default behavior (in the absence of one of these values) is to continue
    emitting whatever version of the old HTTP and networking conventions the
    instrumentation was emitting previously.
- Need to maintain (security patching at a minimum) their existing major version
  for at least six months after it starts emitting both sets of conventions.
- May drop the environment variable in their next major version and emit only
  the stable HTTP and networking conventions.

## Summary of changes

This section summarizes the changes made to the HTTP semantic conventions
from
[v1.20.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/semantic_conventions/http.md)
to

> [v1.23.0 (stable)](README.md).

Reviewer 3: ```suggestion
[v1.23.1 (stable)](README.md).
```

| `net.protocol.name` &rarr; `network.protocol.name` | Recommended &rarr; Conditionally required if not `http` and `network.protocol.version` is set |
| `net.protocol.version` &rarr; `network.protocol.version` | &bullet; Examples fixed: `2.0` &rarr; `2` and `3.0` &rarr; `3`<br>&bullet; See note if [migrating from <= v1.19.0](#migrating-from--v1190) |
| `net.sock.family` | Removed |
| `net.sock.peer.addr` &rarr; `network.peer.address` | On HTTP server spans: if `http.client_ip` was unknown, then also `net.sock.peer.addr` &rarr; `client.address`; `client.address` must be provided to sampler |
| `net.sock.peer.port` &rarr; `network.peer.port` | Now captured even if same as `server.port` |
| `net.sock.peer.name` | Removed |
| New: `http.request.method_original` | Only captured when `http.request.method` is `_OTHER` |
| New: `error.type` |  |
<!-- prettier-ignore-end -->

References:

- [Common attributes v1.20.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/semantic_conventions/http.md#common-attributes)

> - [Common attributes v1.23.0 (stable)](http-spans.md#common-attributes)

Reviewer 3: ```suggestion
- [Common attributes v1.23.1 (stable)](http-spans.md#common-attributes)
```

### HTTP client span attributes

<!-- prettier-ignore-start -->
| Change | Comments |
| --- | --- |
| `http.url` &rarr; `url.full` | |
| `http.resend_count` &rarr; `http.request.resend_count` | |
| `net.peer.name` &rarr; `server.address` | |
| `net.peer.port` &rarr; `server.port` | Now captured even when same as default port for scheme |
<!-- prettier-ignore-end -->

References:

- [HTTP client span attributes v1.20.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/semantic_conventions/http.md#http-client)

> - [HTTP client span attributes v1.23.0 (stable)](http-spans.md#http-client)

Reviewer 3: ```suggestion
- [HTTP client span attributes v1.23.1 (stable)](http-spans.md#http-client)
```

### HTTP server span attributes

<!-- prettier-ignore-start -->
| Change | Comments |
| --- | --- |
| `http.route` | No change |
| `http.target` &rarr; `url.path` and `url.query` | Split into two separate attributes |
| `http.scheme` &rarr; `url.scheme` | Now factors in [X-Forwarded-Proto][], [Forwarded#proto][] headers |
| `http.client_ip` &rarr; `client.address` | If `http.client_ip` was unknown (i.e., no [X-Forwarded-For][], [Forwarded#for][] headers), then `net.sock.peer.addr` &rarr; `client.address`; now must be provided to sampler |
| `net.host.name` &rarr; `server.address` | Now based only on [Host][Host header], [:authority][HTTP/2 authority], [X-Forwarded-Host][], [Forwarded#host][] headers |
| `net.host.port` &rarr; `server.port` | Now based only on [Host][Host header], [:authority][HTTP/2 authority], [X-Forwarded-Host][X-Forwarded-Host], [Forwarded#host][] headers |
<!-- prettier-ignore-end -->

References:

- [HTTP server span attributes v1.20.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/semantic_conventions/http.md#http-server)

> - [HTTP server span attributes v1.23.0 (stable)](http-spans.md#http-server)

Reviewer 3: ```suggestion
- [HTTP server span attributes v1.23.1 (stable)](http-spans.md#http-server)
```

### HTTP client and server span names

The `{http.method}` portion of span names is replace by `HTTP` when
`{http.method}` is `_OTHER`.

See note if [migrating from `<= v1.17.0`](#migrating-from--v1170).

References:

- [HTTP client and server span names v1.20.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/semantic_conventions/http.md#name)

> - [HTTP client and server span names v1.23.0 (stable)](http-spans.md#name)

Reviewer 3: ```suggestion
- [HTTP client and server span names v1.23.1 (stable)](http-spans.md#name)
```

- **Histogram buckets**: boundaries updated to reflect change from milliseconds
  to seconds, and zero bucket boundary removed
- **Attributes**: see table below

<!-- prettier-ignore-start -->
| Attribute change | Comments |
| --- | --- |
| `http.method` &rarr; `http.request.method` | Now captures only 9 common HTTP methods by default plus `_OTHER` |
| `http.status_code` &rarr; `http.response.status_code` |  |
| `net.peer.name` &rarr; `server.address` |  |
| `net.peer.port` &rarr; `server.port` | Now captured even when same as default port for scheme |
| `net.sock.peer.addr` | Removed |
| `net.protocol.name` &rarr; `network.protocol.name` | Recommended &rarr; Conditionally required if not `http` and `network.protocol.version` is set |
| `net.protocol.version` &rarr; `network.protocol.version` | Examples fixed: `2.0` &rarr; `2` and `3.0` &rarr; `3`; see note if [migrating from `<= v1.19.0`](#migrating-from--v1190) |
| New: `error.type` |  |
<!-- prettier-ignore-end -->

References:

- [Metric `http.client.duration` v1.20.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/metrics/semantic_conventions/http-metrics.md#metric-httpclientduration)

> - [Metric `http.client.request.duration` v1.23.0 (stable)](http-metrics.md#metric-httpclientrequestduration)

Reviewer 3: ```suggestion
- [Metric `http.client.request.duration` v1.23.1 (stable)](https://github.com/open-telemetry/semantic-conventions/blob/v1.23.1/docs/http/http-metrics.md#metric-httpserverrequestduration)
```

| `net.protocol.version` → `network.protocol.version` | Examples fixed: `2.0` → `2` and `3.0` → `3`; see note if [migrating from `<= v1.19.0`](#migrating-from--v1190) |
| `net.host.name` → `server.address` | &bullet; Recommended → Opt-In (due to high-cardinality vulnerability since based on HTTP headers)<br>&bullet; Now based only on [`Host` span][Host header], [`:authority` span][HTTP/2 authority], [`X-Forwarded-Host` span][X-Forwarded-Host], [`Forwarded#host` span][Forwarded#host] headers |
| `net.host.port` → `server.port` | &bullet; Recommended → Opt-In (due to high-cardinality vulnerability since based on HTTP headers)<br>&bullet; Now based only on [`Host` span][Host header], [`:authority` span][HTTP/2 authority], [`X-Forwarded-Host` span][X-Forwarded-Host], [`Forwarded#host` span][Forwarded#host] headers |
| New: `error.type` |  |
<!-- prettier-ignore-end -->

References:

- [Metric `http.server.duration` v1.20.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/metrics/semantic_conventions/http-metrics.md#metric-httpserverduration)

> - [Metric `http.server.request.duration` v1.23.0 (stable)](http-metrics.md#metric-httpserverrequestduration)

Reviewer 3: ```suggestion
- [Metric `http.server.request.duration` v1.23.1 (stable)](https://github.com/open-telemetry/semantic-conventions/blob/v1.23.1/docs/http/http-metrics.md#metric-httpserverrequestduration)
```

## Migrating from a version earlier than v1.20.0?

> 

Reviewer 3: I think it would be helpful to add a quick note that below points are in additional to things listed above
